File: scripts/python/test_utils.py
"""
Shared utilities for Python test log generation scripts.
"""

import argparse
import os
import random

API_URL = os.environ.get("API_URL", "http://localhost:4000/api/logs")

DEFAULT_CHANNELS = ["Moody Blues", "bumblebee", "jazz"]

NAMESPACES = ["auth", "api", "db", "cache", "worker", "scheduler"]

LEVEL_NAMES = ["debug", "info", "warning", "error", "critical"]

MESSAGES = [
    "Processing request",
    "Database query completed",
    "User authenticated",
    "Cache hit",
    "Cache miss",
    "Connection established",
    "Task completed successfully",
    "Retrying operation",
    "Configuration loaded",
    "Service started",
    "Webhook received",
    "Email sent",
    "File uploaded",
    "Payment processed",
    "Notification dispatched",
]

USER_IDS = ["user_001", "user_002", "user_003", "user_admin", "user_guest"]


def random_element(arr: list) -> any:
    """Return a random element from the array."""
    return random.choice(arr)


def random_request_id() -> str:
    """Generate a random request ID."""
    return f"req_{''.join(random.choices('abcdefghijklmnopqrstuvwxyz0123456789', k=8))}"


def random_duration() -> int:
    """Generate a random duration in milliseconds."""
    return random.randint(0, 2000)


def generate_random_extras(level: str) -> dict:
    """Generate random extra fields for a log entry."""
    extras = {}

    if random.random() > 0.5:
        extras["userId"] = random_element(USER_IDS)
    if random.random() > 0.5:
        extras["requestId"] = random_request_id()
    if random.random() > 0.5:
        extras["duration"] = random_duration()
    if random.random() > 0.7:
        extras["metadata"] = {
            "version": "1.0.0",
            "environment": random_element(["dev", "staging", "production"]),
            "region": random_element(["us-east-1", "eu-west-1", "ap-south-1"]),
        }
    if level in ("error", "critical") and random.random() > 0.3:
        extras["error"] = {
            "message": "Something went wrong",
            "code": random_element(
                ["ERR_TIMEOUT", "ERR_NOT_FOUND", "ERR_UNAUTHORIZED", "ERR_INTERNAL"]
            ),
            "stack": "Error: Something went wrong\n    at process (/app/index.py:42)",
        }

    return extras


def parse_args() -> argparse.Namespace:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(description="Insert test logs into abbacchio")
    parser.add_argument(
        "--count", "-c", type=int, default=5, help="Number of logs per channel"
    )
    parser.add_argument(
        "--delay", "-d", type=int, default=100, help="Delay between logs in ms"
    )
    parser.add_argument("--name", "-n", type=str, help="Log name/namespace")
    parser.add_argument(
        "--channel",
        "-C",
        type=str,
        help="Channel name(s), comma-separated",
    )
    return parser.parse_args()


def get_channels(args: argparse.Namespace) -> list[str]:
    """Get channels from args or use defaults."""
    if args.channel:
        return [c.strip() for c in args.channel.split(",")]
    return DEFAULT_CHANNELS


def print_config(args: argparse.Namespace, channels: list[str]) -> None:
    """Print the current configuration."""
    print(f"Inserting {args.count} logs per channel ({', '.join(channels)})")
    print(f"Delay between logs: {args.delay}ms")
    print(f"API URL: {API_URL}")
    print(f"Name: {args.name or 'random'}\n")

File: scripts/js/test-utils.ts
/**
 * Shared utilities for test log generation scripts
 */

import { parseArgs } from 'node:util';

export const API_URL = process.env.API_URL || 'http://localhost:4000/api/logs';

export const defaultChannels = ['API', 'Worker1', 'Worker2','Worker3','Worker4','Worker5','Worker6','Worker7'] as const;

export const namespaces = ['init', 'processing', 'shutdown', 'cache', 'worker', 'scheduler'];

export const levelNames = ['trace', 'debug', 'info', 'warn', 'error', 'fatal'] as const;

export const messages = [
  'Processing request',
  'Database query completed',
  'User authenticated',
  'Cache hit',
  'Cache miss',
  'Connection established',
  'Task completed successfully',
  'Retrying operation',
  'Configuration loaded',
  'Service started',
  'Webhook received',
  'Email sent',
  'File uploaded',
  'Payment processed',
  'Notification dispatched',
];

export const userIds = ['user_001', 'user_002', 'user_003', 'user_admin', 'user_guest'];

export function randomElement<T>(arr: readonly T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

export function randomRequestId(): string {
  return `req_${Math.random().toString(36).substring(2, 10)}`;
}

export function randomDuration(): number {
  return Math.floor(Math.random() * 2000);
}

export function generateRandomExtras(level: string): Record<string, unknown> {
  const extras: Record<string, unknown> = {};

  if (Math.random() > 0.5) {
    extras.userId = randomElement(userIds);
  }
  if (Math.random() > 0.5) {
    extras.requestId = randomRequestId();
  }
  if (Math.random() > 0.5) {
    extras.duration = randomDuration();
  }
  if (Math.random() > 0.7) {
    extras.metadata = {
      version: '1.0.0',
      environment: randomElement(['dev', 'staging', 'production']),
      region: randomElement(['us-east-1', 'eu-west-1', 'ap-south-1']),
    };
  }
  if ((level === 'error' || level === 'fatal') && Math.random() > 0.3) {
    extras.error = {
      message: 'Something went wrong',
      code: randomElement(['ERR_TIMEOUT', 'ERR_NOT_FOUND', 'ERR_UNAUTHORIZED', 'ERR_INTERNAL']),
      stack: 'Error: Something went wrong\n    at process (/app/index.js:42:15)',
    };
  }

  return extras;
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export interface ScriptOptions {
  count: number;
  delay: number;
  key?: string;
  name?: string;
  channels: readonly string[];
}

export function parseScriptArgs(defaultCh: readonly string[]): ScriptOptions {
  const { values } = parseArgs({
    options: {
      count: { type: 'string', short: 'c', default: '5' },
      delay: { type: 'string', short: 'd', default: '100' },
      key: { type: 'string', short: 'k' },
      name: { type: 'string', short: 'n' },
      channel: { type: 'string', short: 'C' },
    },
  });

  const channels = values.channel
    ? values.channel.split(',').map((c: string) => c.trim())
    : defaultCh;

  return {
    count: parseInt(values.count!, 10),
    delay: parseInt(values.delay!, 10),
    key: values.key,
    name: values.name,
    channels,
  };
}

export function printConfig(options: ScriptOptions): void {
  console.log(`Inserting ${options.count} logs per channel (${options.channels.join(', ')})`);
  console.log(`Delay between logs: ${options.delay}ms`);
  console.log(`API URL: ${API_URL}`);
  console.log(`Encryption: ${options.key ? 'enabled' : 'disabled'}`);
  console.log(`Name: ${options.name || 'random'}\n`);
}
